hash.go

```diff
@@ -13,10 +13,43 @@ import (
 	"io"
 	"os/exec"
 	"strings"
+	"math/rand"
 )
 
 const buildIDSeparator = "/"
 
+	// hashLength is the number of base64 characters to use for the final
+	// hashed name.
+	// This needs to be long enough to realistically avoid hash collisions,
+	// but short enough to not bloat binary sizes.
+	// The namespace for collisions is generally a single package, since
+	// that's where most hashed names are namespaced to.
+	// Using a "hash collision" formula, and taking a generous estimate of a
+	// package having 10k names, we get the following probabilities.
+	// Most packages will have far fewer names, but some packages are huge,
+	// especially generated ones.
+	// We also have slightly fewer bits in practice, since the base64
+	// charset has 'z' twice, and the first base64 char is coerced into a
+	// valid Go identifier. So we must be conservative.
+	// Remember that base64 stores 6 bits per encoded byte.
+	// The probability numbers are approximated.
+	//
+	//    length (base64) | length (bits) | collision probability
+	//    -------------------------------------------------------
+	//           4               24                   ~95%
+	//           5               30                    ~4%
+	//           6               36                 ~0.07%
+	//           7               42                ~0.001%
+	//           8               48              ~0.00001%
+	//
+	// We want collisions to be practically impossible, so we choose 8 to
+	// end up with a chance of about 1 in a million even when a package has
+	// thousands of obfuscated names.
+	// HashLength has been randomized as an attempt to evade AV signatures.
+
+var hashLength = rand.Intn(28)+8 // 8 <= hashLength <= b64SumBuffer
```

Comment:
The variance between 8 and 36 seems too high - hashes averaging ~25 bytes rather than 8 will significantly bloat the size of obfuscated code and binaries. How about a variance between 8 and 16?

Using a global random integer feels wrong, though. You will end up with the same random hash length for the entire package. But also, this calls `rand.Intn` before we've called `rand.Seed` in the main function, so you're essentially almost always getting the same random number due to the initial default seed of 0. So I think the varying hash length should happen in hashWithCustomSalt.

For example, you could take the _last_ byte of the hash, interpret its lowest three bits as an integer, and use that as a number between 0 and 7 to add to the minimum hash length of 8. That would give us a deterministic number between 8 and 15.

Reply:
Sorry for the very late reply
I just noticed right now that rand does get called before seeding -- blame it on the codeine.

Also pretty on-point for decreasing the variable length range
But I don't see the point of the extra complexity when we determine the size inside hashWithCustomSalt (if I understood you correctly) when we can just fix it determining the hashLength in main and add it as an arg wherever it may be needed, that shouldn't break anything, right?

Reply:
> But I don't see the point of the extra complexity when we determine the size inside hashWithCustomSalt

Right now, the hashed name length is fixed - always 8 characters.

In your patch, it is randomized _per package_. So if a package ends up with a random length of 11, _all the names_ declared in that package will use a hash with that length.

I argue that that's not great, because we could make the randomization _per name_. We just need the obfuscated name `pkg.Foo` to always end up as the same hashed name with the hashed length, but we don't need `pkg.Foo` and `pkg.Bar` to use the same hash length. In fact, all names declared in a package using the same hash length is a pretty simple pattern to recognize.

What I meant by "using the last byte of the hash to determine the length" is something like:

```
const minHashLength = 8
const maxHashLength = 15
const hashLengthRange = maxHashLength - minHashLength

// The last byte of the hash won't be used as part of the name,
// but it is still deterministic and hard to predict,
// so it provides us with useful randomness between 0 and 255.
// We want the number to be between 0 and hashLenthRange-1.
hashLengthRandomness := b64SumBuffer[len(b64SumBuffer)-1] % hashLengthRange

hashLength := minHashLength + hashLengthRandomness
b64Name := b64SumBuffer[:hashLength]
```

I _think_ this would work, though I haven't tested it. You could verify it by adding debug prints to the function, and seeing that each package ends up with names that obfuscate to different lengths. But the tests should still pass.

Reply:
Okay, It does work, Tests are a pass.

However, I've noticed that we cannot use the last byte of b64SumBuffer, since its length is 45 and therefore always padded with a =, it will always just add the same value of 13.
```bash
[hashLengthRandomness] [hashLength] [b64Name] [b64SumBuffer]
5 13 eN86j9GJ27BbQ eN86j9GJ27BbQEph73xzNAriumxHc_RsCXPBlC7W40k=
5 13 WScA1mZVCKQLL WScA1mZVCKQLLuDyijSjlJrfkTKb1GdH8U_8G2I0M8o=
5 13 Djfx_R_ALi2OH Djfx_R_ALi2OH3AB4anogu7YbshjgDE9ll0ra50Ni54=
```
We can fix it by using the second-to-last byte instead of the last: `b64SumBuffer[len(b64SumBuffer)-2]`

```bash
[hashLengthRandomness] [hashLength] [b64Name] [b64SumBuffer]
6 14 4esVcNsb2ook4n 4esVcNsb2ook4ncoThLvXKpeKp7K6BQK2nB1jIuv8a0=
3 11 NuqeKn7PK8M NuqeKn7PK8M2FBT7cpwdxo7biesDn9tpSzMoQo9clE4=
2 10 wuwmUCSCK9 wuwmUCSCK9I9zFcaPPaC4EWAATyV3eQIA9u2vbvCFZk=
0 8 Ve4Nh0_a Ve4Nh0_ak_ZZx5b_u8lKLNEjt08XhNaQI5WnUQGkKmM=
```

Reply:
Oh, right, the padding. If you add `.WithPadding(NoPadding)` to `nameBase64`, then we don't have to worry about that at all, and we can use the last byte as suggested earlier. Your current code makes me slightly uneasy, because if we somehow end up with more padding in the future, we could silently go back to always using a padding byte. Easier to reason about it when there's no padding, and we don't need any padding anyway.

Reply:
Sorry for the late reply
I did think of NoPadding before, however, the quirk with Base64.NoPadding is that it actually delimits the Base64 buffer with a `[]byte(0)`, it doesn't actually clip out the `=`, just transforms it to that.

And of course this causes hashLengthRandomness to always default to 8, since the last byte will always be [ 0 ].

Reply:
Fair point. It doesn't do the zeroing without padding as far as I can tell, but we always base64-encode the same amount of bytes, so the remaining are left as the initial null bytes.

I'll merge this for now, and post a follow-up PR to try to improve the code a bit. Thanks again!